src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/AlwaysEncrypted/ColumnMasterKeyMetadata.cs

```diff
@@ -0,0 +1,114 @@
+/// </item>
+/// </list>
+/// <para>
+/// This takes ownership of the RSA instance supplied to it, disposing of it when Dispose is called.
+/// </para>
+/// </remarks>
+internal readonly ref struct ColumnMasterKeyMetadata // : IDisposable
+{
+    private static readonly HashAlgorithmName s_hashAlgorithm = HashAlgorithmName.SHA256;
+
+#if NET
+    [InlineArray(SHA256.HashSizeInBytes)]
+    private struct Sha256Hash
+    {
+        private byte _elementTemplate;
+    }
+
+    private readonly Sha256Hash _hash;
+#else
+    private readonly byte[] _hash;
+#endif
+    private readonly RSA _rsa;
+
+    // @TODO: SqlColumnEncryptionCertificateStoreProvider.SignMasterKeyMetadata and .VerifyMasterKeyMetadata should use this type.
+    public ColumnMasterKeyMetadata(RSA rsa, string masterKeyPath, string providerName, bool allowEnclaveComputations)
+    {
+        // Lay the column master key metadata out in memory. Then, calculate the hash of this metadata ready for signature or verification.
+        // .NET Core supports Spans in more places, allowing us to allocate on the stack for better performance. It also supports the
+        // SHA256.HashData method, which saves allocations compared to instantiating a SHA256 object and calling TransformFinalBlock.
+
+        // By this point, we know that we have a valid certificate, so the path is valid. The longest valid masterKeyPath is in the format:
+        // [LocalMachine|CurrentUser]/My/[40 character SHA1 thumbprint]
+        // ProviderName is a constant string of length 23 characters, and allowEnclaveComputations' longest value is 5 characters long. This
+        // implies a maximum length of 84 characters for the masterKeyMetadata string - and by extension, 168 bytes for the Unicode-encoded
+        // byte array. This is small enough to allocate on the stack, but we fall back to allocating a new char/byte array in case those assumptions fail.
+        ReadOnlySpan<char> enclaveComputationSpan = (allowEnclaveComputations ? bool.TrueString : bool.FalseString).AsSpan();
+        int masterKeyMetadataLength = providerName.Length + masterKeyPath.Length + enclaveComputationSpan.Length;
+        int byteCount;
+
+#if NET
+        const int CharStackAllocationThreshold = 128;
+        const int ByteStackAllocationThreshold = CharStackAllocationThreshold * sizeof(char);
+
+        Span<char> masterKeyMetadata = masterKeyMetadataLength <= CharStackAllocationThreshold
+            ? stackalloc char[CharStackAllocationThreshold].Slice(0, masterKeyMetadataLength)
+            : new char[masterKeyMetadataLength];
+        Span<char> masterKeyMetadataSpan = masterKeyMetadata;
+#else
+        char[] masterKeyMetadata = new char[masterKeyMetadataLength];
+        Span<char> masterKeyMetadataSpan = masterKeyMetadata.AsSpan();
+#endif
+
+        providerName.AsSpan().ToLowerInvariant(masterKeyMetadataSpan);
+        masterKeyPath.AsSpan().ToLowerInvariant(masterKeyMetadataSpan.Slice(providerName.Length));
```

Comment:
Are the characters in the input variable restricted such that it is certain that the output lower invariant replacement version is certain to be of exactly the same length?

Reply:
Yes, indirectly. This is due to be used by the certificate store provider, and the constructor takes an `RSA` instance. The only way to get this instance is by using the master key path to select a valid certificate, and the only valid master key paths are ASCII text in a few formats:
* `[LocalMachine|CurrentUser]/My/[40 character SHA1 thumbprint]`
* `My/[40 character SHA1 thumbprint]`
* `[40 character SHA1 thumbprint]`

Reply:
Excellent. It might be worth leaving a comment in the code somewhere so anyone who reads the code in future knows that there are input restrictions.